### ISSUE TITLE
Show/hide sidebar, hide/show editor and put the content of sidebar in focus/put the content of the editor in focus in one single shortcut

### ISSUE THREAD

**Issue by ralf3u** [xiota/geany-preview#53]
Today I asked myself why I don't use so much the plugin Preview any more like before. Suddenly I had the idea that one important shortcut is missing (see title of this issue).

What do you think about that?

**Comment by xiota** [comment]
If I recall, there's a lua script that can change views: editor → editor+sidebar → sidebar

Is the main difference between this issue and the lua script where focus goes?

Previous attempts to focus sidebar contents has not been (consistently) successful, so it may not be doable.

**Comment by ralf3u** [comment]
>If I recall, there's a lua script that can change views: editor → editor+sidebar → sidebar

I didn't feel comfortable with the shortcut `editor → editor+sidebar → sidebar`. But till today I couldn't describe what could be better.
The new shortcut would be simply `editor ↔ sidebar`. Because at the moment I don't use the auto-sidebar, this shortcut would be just perfect.

>Is the main difference between this issue and the lua script where focus goes?

No, in the lua script the focus should be in the content of the sidebar, when only the sidebar appears, but it does not work because of lua script.
The main difference between this issue and the lua script is that there is no step between. In this new issue this middle step does not exist: `editor+sidebar`.

>Previous attempts to focus sidebar contents has not been (consistently) successful, so it may not be doable.

It didn't work because of lua script. If this issue would be programed directly in the plugin Preview, then I think that the focus should work correctly because in the plugin Preview the focus works (see #4). 

Some hours ago I had the idea, that, when programing the new shortcut, the order of the new issue could be changed like this in case the new shortcut does not work:
Show/hide sidebar, put the content of the sidebar in focus/put the content of the editor in focus and hide/show editor.

**Comment by xiota** [comment]
So it would be similar to this lua script, but with focus? [`switch-views-(editor-↔-sidebar).lua`](https://github.com/xiota/geanylua-scripts/blob/main/switch-views/switch-views-(editor-%E2%86%94-sidebar).lua)

I kind of want to rewrite the preview plugin... don't know when/if I'll get to it though.

**Comment by ralf3u** [comment]
Sorry, I just recognized that the shortcut `editor ↔ sidebar` does already exist in the switch views. It would be better if this shortcut would be directly in the plugin Preview to solve the focus.

Anyway the shortcut `editor ↔ sidebar` does not work perfect. If auto-sidebar is not active, and if Geany is not open, and if I open an HTML-document, then I have to press the first time 2x the shortcut to make the shortcut work. Afterwards the shortcut works correctly. So, every time the shortcut is pressed 1x, there is a toggle.

**Comment by ralf3u** [comment]
>So it would be similar to this lua script, but with focus?
editor↔sidebar

Exactly. The really big thing what I miss directly in the plugin Preview is the shortcut `editor ↔ sidebar`, where the focus problem is solved.

>I kind of want to rewrite the preview plugin... don't know when/if I'll get to it though.

There is no hurry.

**Comment by ralf3u** [comment]
>Anyway the shortcut editor ↔ sidebar does not work perfect. If auto-sidebar is not active, and if Geany is not open, and if I open an HTML-document, then I have to press the first time 2x the shortcut to make the shortcut work. Afterwards the shortcut works correctly. So, every time the shortcut is pressed 1x, there is a toggle.

It is not the only problem with the lua shortcut `editor ↔ sidebar`. I just added in geanylua only the switch-views, so no auto-sidebar and no toggle-menu-bar. When in Geany the sidebar is not visible and if I press the shortcut `editor ↔ sidebar` multiple times, the sidebar never appears.

**Comment by ralf3u** [comment]
Once the shortcut `editor↔sidebar` is part of the plugin Preview, I'm sure that I will not use the lua script switch-views any more. Only the two lua scripts will be still interesting for me: auto-sidebar (only when using a big screen) and toggle-menu-bar.

**Comment by xiota** [comment]
I would prefer not to add general shortcuts like that to the plugin, but will think about it.  Ubuntu 24.10 should be released in a couple months, so that gives a time frame to decide.

**Comment by ralf3u** [comment]
The current lua script shortcut `editor ↔ sidebar` has three issues:
- If auto-sidebar is not active, and if Geany is not open, and if one opens a HTML-document from a file-manager, then one need to press the first time 2x the shortcut to make the shortcut work.
- When in Geany the sidebar is not visible and if one press the shortcut `editor ↔ sidebar` 2x or multiple times, then the sidebar never appears.
- Focus does not work in the content area of the sidebar.

Focus does not work in lua. Focus works in the plugin Preview when using a shortcut.
Remark: Focus does not work in the plugin Preview only when using the mouse for the tab-title (see #15).

Some weeks ago I started to code a website. For this I deactivated the plugin Preview in the list of plugins because I struggled with the issues of the shortcuts. I was myself astonished about my decision to deactivate the plugin Preview. After the creation of the website, I asked myself what were the reasons that I deactivated the plugin Preview and how should the plugin be changed that I don't deactivate it in the future. I recognized that the shortcut `editor ↔ sidebar` is a very important element of Geany Preview and it should work perfect. If this shortcut would work perfect then I would not deactivate the plugin Preview.

>I would prefer not to add general shortcuts like that to the plugin, [...]

The shortcut `editor ↔ sidebar` that also can be named as `Toggle between editor and sidebar` would be the second shortcut and at the same time the last shortcut in the plugin Preview. The first shortcut `Toggle focus between editor and sidebar` works already perfect.
An alternative would be to put both shortcuts in an other plugin.

If the shortcut `Toggle between editor and sidebar` would work in the plugin Preview then there are for me two different modes how I would use the plugin Preview:
- enable sidebar: I would use this when using a wide screen. On small screens I would use this mode only for the creation of layout. The interesting shortcut for this mode is `Toggle focus between editor and sidebar`.
- disable sidebar: I would use this mode specially on small screens to edit text. The interesting shortcut for this mode is `Toggle between editor and sidebar`. I think that this mode will be the mode that I will use the most of the time.

I can easily switch between the two modes just by pressing the shortcut `Toggle Sidebar`.

This means that the auto-sidebar small edition can be used. With "small edition" I mean the very first program of auto-sidebar. The auto-sidebar can be useful when sidebar is enabled.

I would not use any more the shortcuts of switch-views. The shortcuts of switch-views are still somehow interesting, but because I would use only the two modes I don't need the other shortcuts of switch-views.

**Comment by ralf3u** [comment]
This would be my dream-team-shortcuts:
F2: `Toggle focus between editor and sidebar`
F3: `Toggle sidebar` (to switch the mode)
F4: `Toggle between editor and sidebar`

There is only one thing to keep in mind when switching mode:
Switch mode **ONLY** when the editor is visible.

Else F3 will show an empty content-area.

I have two questions:
Is it possible to program a shortcut with a IF-THEN-ElSE-instruction?
If it is possible, is it too complicated to program a IF-THEN-ElSE-instruction?

**Comment by ralf3u** [comment]
If the shorcuts are outside the plugin `Preview`, then maybe the name of the new plugin could be `Preview shortcuts`.

There is a difference between the shortcut lua script `editor ↔ sidebar` and the fictional shortcut `Toggle between editor and sidebar`. The starting point of the two shortcuts are different:
- The lua-script starts when editor and sidebar are side by side.
- The starting point of the fictional shortcut `Toggle between editor and sidebar` is when the sidebar is disabled.

If there is the possibility to program a shortcut with an IF-THEN-ELSE-instruction, then those two additional shortcuts as option could be interesting for the very far future:
- `Toggle focus between editor and sidebar` and `Toggle between editor and sidebar` in one single shortcut (IF editor & sidebar are enabled, THEN execute `Toggle focus between editor and sidebar`, ELSE execute `Toggle between editor and sidebar`); the name of the shortcut could be `Toggle focus between editor and sidebar, or Toggle between editor and sidebar`.
- `Switch mode` makes sense because with the shortcut `Toggle sidebar` one have to keep in mind that one can switch the mode without problems only when the editor is visible, else there is an empty content-area when leaving mode 2.

**Comment by xiota** [comment]
>  Is it possible to program a shortcut with a IF-THEN-ElSE-instruction?

Configurable by the user?  It's possible, you can do it with the lua plugin.  I'm very unlikely to add a full scripting language to the preview plugin though.

Modes are usually difficult to program and use.  User has to know what the modes are, remember how they're used, keep track of the current mode, etc.

I plan to look at this sometime this weekend or next week.

**Comment by ralf3u** [comment]
With the shortcut name `Switch mode` I mean this instruction:
IF editor&sidebar are side by side, THEN disable sidebar; ELSE show editor&sidebar side by side.

>Modes are usually difficult to program and use. User has to know what the modes are, remember how they're used, keep track of the current mode, etc.

Yes, you are right. So, an alternative for the shortcut name `Switch mode` could be:
`Switch from editor&sidebar to editor, or switch from editor or sidebar to editor&sidebar`.

So this would the dream-team-shortcuts:
F2: `Toggle focus between editor and sidebar`
F3: `Switch from editor&sidebar to editor, or switch from editor or sidebar to editor&sidebar`
F4: `Toggle between editor and sidebar`

**Comment by ralf3u** [comment]
If the shortcut `Switch from editor&sidebar to editor, or switch from editor or sidebar to editor&sidebar` is pressed, the result should always be: The content area of the editor is in focus, so a blinking cursor in the editor.

**Comment by ralf3u** [comment]
The shortcut `Switch from editor&sidebar to editor, or switch from editor or sidebar to editor&sidebar` could be programed with lua, because the focus works there correctly. I just tested those both shortcuts because the shortcut `Switch from editor&sidebar to editor, or switch from editor or sidebar to editor&sidebar` is somehow a combination of both shortcuts:
- `Show both editor and sidebar`
- `Switch views (editor ↔ sidebar)`

But unfortunately the shortcut `Toggle between editor and sidebar` can't be programed in lua because there the focus does not work correctly in the sidebar (the focus is not in the content area of the sidebar).

**Comment by ralf3u** [comment]
Same problem for this shortcut:
`Toggle focus between editor and sidebar, or Toggle between editor and sidebar` (see https://github.com/xiota/geany-preview/issues/53#issuecomment-2344408243). This shortcut can't be programed with lua because of the focus issue.

**Comment by ralf3u** [comment]
To avoid misunderstanding, I would like to repeat things in a different way.

If the shortcut `Toggle between editor and sidebar` would exist, then this would be my dream-team-shortcuts:
F2: `Toggle focus between editor and sidebar`
F3: `Toggle sidebar`
F4: `Toggle between editor and sidebar`
There is only one thing that I have to keep in mind when switching mode with F3:
Switch mode ONLY when the editor is visible, else F3 will show an empty content-area when leaving mode 2.

If the shortcut `Toggle between editor and sidebar` and the shortcut `Switch from editor&sidebar to editor, or switch from editor or sidebar to editor&sidebar` would exist, then this would be my dream-team-shortcuts:
F2: `Toggle focus between editor and sidebar`
F3: `Switch from editor&sidebar to editor, or switch from editor or sidebar to editor&sidebar`
F4: `Toggle between editor and sidebar`

If the shortcut `Switch from editor&sidebar to editor, or switch from editor or sidebar to editor&sidebar` and the shortcut `Toggle focus between editor and sidebar, or toggle between editor and sidebar` would exist, then this would be my dream-team-shortcuts:
F2: `Toggle focus between editor and sidebar, or toggle between editor and sidebar`
F3: `Switch from editor&sidebar to editor, or switch from editor or sidebar to editor&sidebar`

The last dream-team-shortcut is the ultimative dream-team-shortcut because there are only two shortcuts instead of three shortcuts. The help to remember the two shortcuts is to keep in mind that F3 is a kind of Switch-mode-shortcut. One can not reduce the shortcuts even more.

If all the shortcuts from above would exist, then this is the list in the Preferences in Keybindings:
`Toggle focus between editor and sidebar`
`Toggle between editor and sidebar`
`Toggle focus between editor and sidebar, or toggle between editor and sidebar`
`Switch from editor&sidebar to editor, or switch from editor or sidebar to editor&sidebar`

Only the last shortcut can be programed in lua because the other shortcuts in lua would create a focus issue. If this shortcut will be programed in lua, then it would be better not to add it to the other shortcuts in switch-views, first because I don't need the shortcuts of switch-views any more, and second because the auto-sidebar small edition should work.

**Comment by ralf3u** [comment]
If a user wants to use only mode 1, so editor&sidebar side by side, then there is the possibility during mode 1 to see only the editor by pressing the shortcut `Toggle sidebar.` But If the user wants to see during mode 1 only the sidebar there is no shortcut in the list of my last comment. So the shortcut would be: `Toggle between editor&sidebar and sidebar`. Also this can't be programed correctly with lua because of the focus issue.

This is the whole list of the shortcuts:
`Toggle focus between editor and sidebar`
`Toggle between editor&sidebar and sidebar`
`Toggle between editor and sidebar`
`Toggle focus between editor and sidebar, or toggle between editor and sidebar`
`Switch from editor&sidebar to editor, or switch from editor or sidebar to editor&sidebar`

Only the last shortcut can be programed with lua without focus issue, I think so.

**Comment by ralf3u** [comment]
I would like to change the name of some shortcuts of my last comments to understand better what the shortcuts are doing. This is the whole list of shortcuts – I added issue number 55 to the list:

`Copy`
`Toggle focus between editor and sidebar`
`Toggle between editor&sidebar and only sidebar`
`Toggle between only editor and only sidebar`
`If editor&sidebar are shown then toggle focus between editor and sidebar, else toggle between only editor and only sidebar`
`Switch from editor&sidebar to only editor, otherwise switch to editor&sidebar`

**Comment by ralf3u** [comment]
I recognized that the two shortcuts `Toggle focus between editor and sidebar` and `Toggle between only editor and only sidebar` can be removed from the list because it is part of the shortcut `If editor&sidebar are shown then toggle focus between editor and sidebar, else toggle between only editor and only sidebar`. I removed also the shortcut `Copy` from the list because it does not belong to the moving actions. So, this is the whole list of shortcuts:

`Toggle between editor&sidebar and only sidebar`
`If editor&sidebar are shown then toggle focus between editor and sidebar, else toggle between only editor and only sidebar`
`Switch from editor&sidebar to only editor, otherwise switch to editor&sidebar`

Now there are only three shortcuts. This is user-friendly.

Only the last shortcut can be programed with Lua because the focus works correct there. The shortcut stands for "change mode". For that one could use the existing shortcut `Toggle Sidebar`, but the problem is: If only the sidebar is visible, and if the shortcut `Toggle Sidebar` is pressed, then the content-area of the tab is empty. A help is to remember to press the shortcut `Toggle Sidebar` only if the editor is visible, is it only editor or editor&sidebar.

**Comment by ralf3u** [comment]
If the three shortcuts of my last comment would exist, I would set them to F2, F3 and F4.

F2 (it would be my most used shortcut in preview):
`If editor&sidebar are shown then toggle focus between editor and sidebar, else toggle between only editor and only sidebar`

F3 (it would be my second most used shortcut in preview; it stands for "change mode")
`Switch from editor&sidebar to only editor, otherwise switch to editor&sidebar`

F4
`Toggle between editor&sidebar and only sidebar`

**Comment by ralf3u** [comment]
I recognized that two names of the shortcuts are too long because the dialog-window of the `Preferences` becomes too long. So, I need to make them shorter. At the moment this is the longest name in geanylua:
switch-views-(editor-→-editor+sidebar-→-sidebar).lua
The length is OK.

This is the list of shortcuts from https://github.com/xiota/geany-preview/issues/53#issuecomment-2380877522:
`Toggle between editor&sidebar and only sidebar`
`If editor&sidebar are shown then toggle focus between editor and sidebar, else toggle between only editor and only sidebar`
`Switch from editor&sidebar to only editor, otherwise switch to editor&sidebar`

I changed two names. And I changed the order. But all those changes do not change the functionality of the shortcuts. So, this is the new list:
`Toggle between editor&sidebar and only sidebar`
`Toggle between editor&sidebar and only editor`
`Toggle (focus) or (only editor / only sidebar)`

The new names sound somehow perfect.
